Sources/Networking/HTTPClient.swift

```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
+                let result: [String: Any]?
+
+                do {
+                    result = try JSONSerialization.jsonObject(with: data) as? [String: Any]
+                } catch let jsonError {
+                    Logger.error(Strings.network.parsing_json_error(error: jsonError))
+
+                    let dataAsString = String(data: data, encoding: .utf8) ?? ""
+                    Logger.error(Strings.network.json_data_received(dataString: dataAsString))
+
+                    throw jsonError
                 }
 
-                httpResponse = self.eTagManager.httpResultFromCacheOrBackend(with: httpURLResponse,
-                                                                             jsonObject: jsonObject,
-                                                                             error: receivedJSONError,
-                                                                             request: urlRequest,
-                                                                             retried: request.retried)
-                if httpResponse == nil {
-                    Logger.debug(Strings.network.retrying_request(httpMethod: request.method.httpMethod,
-                                                                  path: request.path))
-                    self.state.modify {
-                        $0.queuedRequests.insert(request.retriedRequest(), at: 0)
-                    }
+                guard let result = result else {
+                    throw ErrorUtils.unexpectedBackendResponseError()
                 }
+
+                return result
+            } else {
+                return [:]
             }
         }
 
-        var networkError = networkError
-        if dnsChecker.isBlockedAPIError(networkError),
-           let blockedError = dnsChecker.errorWithBlockedHostFromError(networkError) {
-            Logger.error(blockedError.description)
-            networkError = blockedError
-        }
+        let statusCode = HTTPStatusCode(rawValue: httpURLResponse.statusCode)
+
+        Logger.debug(Strings.network.api_request_completed(request.httpRequest,
+                                                           httpCode: statusCode))
+
+        let result: Result<HTTPResponse?, Error> = Result { try extractBody(httpURLResponse, statusCode) }
+            .map { HTTPResponse(statusCode: statusCode, jsonObject: $0) }
+            .map {
+                return self.eTagManager.httpResultFromCacheOrBackend(with: httpURLResponse,
+                                                                     jsonObject: $0.jsonObject,
+                                                                     request: urlRequest,
+                                                                     retried: request.retried)
+            }
+
+        let shouldRetry: Bool = {
+            if case .success(.none) = result {
+                return true
+            } else {
+                return false
+            }
+        }()
```

Comment:
it took me a bit to figure out that this means that we should retry if we got an empty but successful result. 
I think this is equivalent, right? Using our Result extensions
```swift
let shouldRetry = result.error == nil && result.value == nil
```

Reply:
I agree, I hated this too. I replaced how this is done in #1431 with a `Result` extension.